### ISSUE TITLE
Idiomatic way to support Function Call tags

### ISSUE THREAD

**Issue by farzadshafiee** [idehub/react-native-google-analytics-bridge#256]
I want to leverage [Function Call](https://developers.google.com/tag-manager/ios/v3/#function-call-tag) to streamline a few of our integrations and bypass app update cylces with every configuration change. This is a powerful feature of GTM can be applied for many use-cases.

I was looking into the possibility of registering functions in JS realm, but based on the Google documentations, the function must be registered in native realm. I think it might be a good idea to have at least one global function tag handler registered with the container by default, which when executed, passes the arguments to a counterpart callback in the JS realm and following becomes possible:

```js
import { GoogleTagManage } from 'react-native-google-analytics-bridge'
GoogleTagManager.handleFunctionTag(args => {
  // Process the tag...
})
```

The downside of this compared to native integration is that only one _function name_ (which can be configurable) is available to be used when setting up the tags in Tag Manager interface, but prevents `GoogleTagManager.openContainerWithId` method to become obselete when a **Function Call** tag is to be used.

Would you accept PRs for this?

**Comment by cbrevik** [comment]
Hi @farzadshafiee 

This seems like an interesting feature. Would definitely accept PRs. But lets discuss this, I have a couple of questions first.

> The downside of this compared to native integration is that only one function name (which can be configurable) is available to be used when setting up the tags in Tag Manager interface

Are you sure about this? From the code it seems like you can arbitrarily use keys:
```obj-c
  [self.container registerFunctionCallTagHandler:[[MyFunctionCallTagHandler alloc] init]
                                          forTag:kMyTagFunctionName];

```
Seems to me that `kMyTagFunctionName` here could instead be a string value passed from the JavaScript side instead. The fact that the native class `MyFunctionCallTagHandler` handles the function call natively isn't important it seems, all it needs to do in turn is send an event over the JS<->Native bridge with the string value. Or am I misunderstanding something? 

> [..] but prevents GoogleTagManager.openContainerWithId method to become obselete when a Function Call tag is to be used.

Not sure I understand this sentence. What do you mean?

Also, is this possible for Android as well? Can't seem to find docs for this features in v4 of the Google Analytics Android SDK.

**Comment by farzadshafiee** [comment]
First of all, sorry about the duplicate issue. Not sure what happened there.

> Are you sure about this? From the code it seems like you can arbitrarily use keys

You are right. Don't why but I thought the name should match a function inside the native code. All the better...

> Not sure I understand this sentence. What do you mean?

With the current implementation, since there's no way to register a function, anyone who wants to use this specific type of tag, should open the container in native environment and pass that container. The `openContainerWithId` JS method becomes useless for the containers that include such Tags. I was just providing my reasoning behind my proposal to include this API in the package.

> Also, is this possible for Android as well? Can't seem to find docs for this features in v4 of the Google Analytics Android SDK.

The documentation [Google Tag Manager for Android V4](https://developers.google.com/tag-manager/android/v4/#intro) (not Analytics) mentions the support (look for **Custom Function Call**), but the actual documentation lives on the [V3 reference](https://developers.google.com/tag-manager/android/v3/#function-call-tag).

**Comment by farzadshafiee** [comment]
I'll work on the PR if you don't mind.

**Comment by cbrevik** [comment]
Sure, go ahead! I only ask that you do both iOS and Android (since supporting both is a goal of this lib).

Tip for sending events to JS: https://facebook.github.io/react-native/docs/native-modules-ios.html#sending-events-to-javascript

**Comment by farzadshafiee** [comment]
@cbrevik, there were a few drawbacks with RCTEventEmitter on iOS to send the events to JS. Also, as it turns out, implementing Function Call tag for Android v4 is differs from v3 and I had to look through the [API reference](https://developers.google.com/android/reference/com/google/android/gms/tagmanager/Container.FunctionCallTagCallback) itself to find the proper classes and methods. Anyhow, I was finally able to get it working on both platforms. I tried to comment my way throughout the code to reduce confusion about the few decisions I had make.
I was also not sure about your coding style and module definition preference. Feel free to change whatever part of it you think is not following your rules.
